---
title: "APOE_finemapping" 
author: "Brian M. Schilder"
date: "Most recent update:<br> `r Sys.Date()`"
output: 
  rmarkdown::html_document: 
    theme: spacelab
    highlight: zenburn 
    code_folding: show 
    toc: true 
    toc_float: true
    smooth_scroll: true
    number_sections: false 
    self_contained: true
---

```{r setup, include=T, message=F, dpi=300}
root.dir <- here::here()
knitr::opts_chunk$set(echo = T, root.dir=root.dir)
knitr::opts_knit$set(root.dir=root.dir)

library(dplyr)
library(ggplot2)

library(echolocatoR)
```

# Import fine-mapping results

## Query fine-mapping portal

Download Alzheimer's Disease GWAS fine-mapping results for the APOE locus via the [echolocatoR Fine-mapping Portal](https://rajlab.shinyapps.io/Fine_Mapping_Shiny/) API.

```{r}
local_finemap <- echolocatoR::GITHUB.portal_query(dataset_types = "GWAS", 
                                                 phenotypes = "Alzheimer", 
                                                 LD_panels = c("UKB", "1KGphase3"),
                                                 loci = "APOE", 
                                                 file_types = "multi_finemap", 
                                                 results_dir = "./data", 
                                                 overwrite = F)

```

## Merge fine-mapping results

```{r}

finemap_dat <- lapply(local_finemap,function(x){
  dataset <- basename(dirname(dirname(dirname(x))))
  LD_ref <- stringr::str_split(basename(x),pattern = "[.]")[[1]][2]
  printer("Importing",dataset)
  dat <- data.table::fread(x) 
  cbind(dataset=dataset, 
        LD_ref=LD_ref, 
        dat)
}) %>% data.table::rbindlist()


```

### Identify credible sets

Show just the Union Credible Set SNPs (SNPs nominated by at least 1/4 fine-mapping tools).

```{r}
ucs_snps <- subset(finemap_dat, Support>0)
createDT(ucs_snps)
```

### Identify Consensus SNPs

Filter to just the consensus SNPs (SNPs nominated by 2 or more fine-mapping tools).

```{r}
consensus_snps <- subset(finemap_dat, Consensus_SNP==T)
createDT(consensus_snps)
```

# Cell-type-specific annotations

## Epigenomic peak overlap

```{r Epigenomic peak overlap, warning=F, message=F, fig.height=4, fig.width=10}

finemap_dat$Gene<- finemap_dat$Locus
finemap_dat$Locus <- paste(finemap_dat$Locus,finemap_dat$dataset,sep="_")
plot_res <- echolocatoR::SUMMARISE.peak_overlap_plot(merged_DT = finemap_dat,
                                                     snp_filter = "Support>0",
                                                     include.NOTT_2019_peaks = T,
                                        include.NOTT_2019_enhancers_promoters = T,
                                        include.NOTT_2019_PLACseq = T,
                                        include.CORCES_2020_scATACpeaks = T,
                                        include.CORCES_2020_Cicero_coaccess = T,
                                        include.CORCES_2020_bulkATACpeaks = T,
                                        include.CORCES_2020_HiChIP_FitHiChIP_coaccess = T,
                                      include.CORCES_2020_gene_annotations = T,
                                      plot_celltype_specificity = T,
                                      plot_celltype_specificity_genes = T)

```

## Show the epigenomic overlap results

```{r}
createDT(plot_res$data)
```

## Summarise assay counts

Summarise the number of assays in which at least one UCS SNP overlapped with a cell-type-specific epigenomic peak.

```{r}
celltype_counts <- plot_res$data %>% 
  dplyr::group_by(Locus, Cell_type) %>%
  dplyr::summarise(Count=sum(Count,na.rm = T)) %>%
  dplyr::arrange(Locus, desc(Count))
createDT(celltype_counts)
```

## SNP-level results

The above reports are useful to see thing in aggregate, but here we look at the SNP-level results.

- First, we extract cell-type-specific regulatory elements identified in either Nott et al. 2019 or Corces et al. 2020, that overlap with fine-mapped union credible set SNPs (Support>0).   

- Next, we separately extract cell-type-specific interactome anchors that overlap with those regulatory elements. These can be used to link a given regulatory element (e.g. enhancer) to a specific gene (or set of genes) that it affects. 

```{r}
gr.gre <- SUMMARISE.peak_overlap(merged_DT=finemap_dat,
                                   snp_filter="Support>0",
                                   include.NOTT_2019_peaks=T,
                                   include.NOTT_2019_enhancers_promoters=T,
                                   include.NOTT_2019_PLACseq=T,
                                   include.CORCES_2020_scATACpeaks=T,
                                   include.CORCES_2020_Cicero_coaccess=T,
                                   include.CORCES_2020_bulkATACpeaks=T,
                                   include.CORCES_2020_HiChIP_FitHiChIP_coaccess=T,
                                   include.CORCES_2020_gene_annotations=T)  
gr.elements <- subset(gr.gre, !Assay %in% c("HiChIP_FitHiChIP","PLAC"))
non_na_cols <- colnames(gr.elements@elementMetadata)[colSums(!is.na(gr.elements@elementMetadata))>0]

gr.anchors <- subset(gr.gre, Assay %in% c("HiChIP_FitHiChIP","PLAC"))

gr.hits <- clean_granges(GRanges_overlap(GenomicRanges::granges(gr.anchors), 
                                         gr.elements[,non_na_cols],
                                         return_merged = F, verbose = T))
gr.hits$width <- GenomicRanges::width(gr.hits)
unique(gr.elements$Assay)
unique(gr.elements$dataset)
```

```{r}
snp_overlap_counts <- data.frame(gr.hits) %>%   
  dplyr::group_by(Locus, SNP, Cell_type, Consensus_SNP, Support) %>%
  dplyr::summarise(Assay_count=dplyr::n_distinct(Assay)) %>%
   dplyr::arrange(desc(Consensus_SNP), desc(Support), desc(Assay_count))
createDT(snp_overlap_counts)
```

## Gene annotation

Now that we have the hits between fine-mapped SNPs and cell-type-specific regulatory elements, we can try linking those regulatory elements to specific genes.

### Fine-mapped SNP overlap with promoters

Find fine-mapped SNPs that fall within regulatory elements (TTS, intron, promoter, promoter-TSS) connected to specific genes (active in one, some, or all cell-types).

```{r}
promoters_all <-  echolocatoR::NOTT_2019.interactome$H3K4me3_around_TSS_annotated_pe

gr.promoters <- GenomicRanges::makeGRangesFromDataFrame(promoters_all,
                                                        seqnames.field = "chr", 
                                                        start.field = "start", 
                                                        end.field = "end", 
                                                        keep.extra.columns = T) 
gr.overlap1 <- clean_granges(GRanges_overlap(gr.hits, gr.promoters))
gr.overlap1$Assay <- "interactome_H3K4me3_around_TSS"
gr.overlap1$Element_old <- gr.overlap1$Element
gr.overlap1$Element <- stringr::str_split(gr.overlap1$Annotation," ", simplify = T)[,1]
```

### Fine-mapped SNP overlap with enhancers linked to promoters

Find fine-mapped SNPs that fall within *enhancers, superenhancers or promoters* that interact with promoters (active in one, some or all cell-types).

```{r}
range_cols <- grep("interactions$", colnames(promoters_all), value = T)
print(paste(length(range_cols),"interacting element columns identified."))
print(range_cols)
dat <- data.table::melt.data.table(data.table::data.table(promoters_all),
                            variable.name = "interaction_type",
                            measure.vars = range_cols) %>% 
  tidyr::separate(col = "value", sep =",|, ", into=paste0("range",1:200))
non_na_cols <- colnames(dat)[colSums(!is.na(dat))>0]

promoter_interactions <- dat %>% 
  dplyr::select(non_na_cols) %>%
  data.table::melt.data.table(measure.vars = grep("^range",colnames(.), value = T)
                              ) %>%
  subset(!is.na(value) & (value!="")) %>%
  tidyr::separate(col = "value", sep="[:]|-", into=c("CHR","START","END")) %>%
  tidyr::separate(col = "interaction_type", sep="_", into=c("Marker","Element","Assay"),remove = F) %>%
  dplyr::mutate(START=as.numeric(START), END=as.numeric(END),
                peak_coordinates=paste0(Chr,":",Start,"-",End)) %>%
  dplyr::select(-Chr,-Start,-End) %>%
  subset(END>=START)

gr.anchor_bins <- GenomicRanges::makeGRangesFromDataFrame(promoter_interactions, 
                                        seqnames.field = "CHR", 
                                        start.field = "START", 
                                        end.field = "END", 
                                        keep.extra.columns = T)
### Convert markers to Celltypes
cell_dict <- list(neurons="NeuN",
                    microglia="PU1",
                    oligo="Olig2",
                    astrocytes="LHX2",
                    periph="peripheral PU1+")
cell_dict_invert <- setNames(names(cell_dict), cell_dict)
gr.anchor_bins$Cell_type_anchor <- cell_dict_invert[gr.anchor_bins$Marker]
### IMPORTANT!: Return the gr.hits object, not the gr.anchor_bins objects
gr.overlap2 <- GRanges_overlap(gr.anchor_bins[,c("Cell_type_anchor","interaction_type","Element","Gene Name","PeakID","Distance to TSS","Nearest PromoterID","Annotation","Detailed Annotation")],
                                             gr.hits[,c("dataset","Locus","SNP","Support","Consensus_SNP","Cell_type","Assay")], 
                                             return_merged = T) %>%
  clean_granges() %>%
  subset(Cell_type==Cell_type_anchor)
## Remove rows that disagree between annotations
# gr.overlap2 <- subset(gr.overlap2, !(Element=="promoters") & (Element.1!="promoter"))
# gr.overlap2$Element_old <- gr.overlap2$Element
# gr.overlap2$Element  <- gr.overlap2$Element.1
# gr.overlap2$Assay <- "interactome_anchor_bin"
```


### Nominate genes  

Nominate causal genes based on the overlap between fine-mapped variants, active regulatory elements, and interactome data.

For robust variant nomination, we find the intersection between the following groups:  

1. Fine-mapped Consensus SNP AD GWAS variants in the APOE locus (from **echolocatoR**).
2. Coordinates of individual cell-type-specific enhancers/promoters (Table S5 tab 1).
3. Coordinates of clusters of enhancers/superenhancers/promoters (Table S5 tab 1).
4. Coordinates of individual cell-type-specific interactome anchors (Table S5 tabs 5-12).
5. Coordinates of individual promoters/promoter-TSS/TSS/introns with PLAC-seq interactions (Table S5 tab 1). 

Table S5 comes from the supplementary material in [Nott et al. 2019](https://science.sciencemag.org/content/366/6469/1134.abstract). 

```{r} 
# gr.top_hits <- GRanges_overlap(gr.anchor_bins, gr.promoters)
gr.anchors$Cell_type_interactome <- gr.anchors$Cell_type
interactome_cols <- c("count","fdr","ClusterSize","ClusterType","ClusterNegLog10P",
                      "Cell_type_interactome","Anchor")
gr.overlap <- GRanges_overlap(gr.anchors[,interactome_cols],
                              c(gr.overlap1, gr.overlap2),
                              return_merged = T) %>%
  unique() %>%
  # Make sure the data overlaps with a celltype in the interactome
  subset(!is.na(Cell_type_interactome)) %>%
  # make sure celltypes are matching
  subset(Cell_type==Cell_type_interactome)


regulatory_overlap <- data.frame(gr.overlap) %>% 
  # subset(Consensus_SNP==T) %>%
   dplyr::select(all_of(c("dataset", "Locus","SNP","Support", "Gene.Name",
                 "Assay","Element","Cell_type", interactome_cols,
                 "Annotation","PeakID","Distance.to.TSS"
                  # grep("promoter|enhancer|interactions",
                  #      colnames(gr.overlap@elementMetadata), value = T)
                  )), -Cell_type_interactome
                )  %>%
  tidyr::separate(col="Locus", into = c("Locus"), sep = "_", extra = "drop") %>%
  unique() %>%
  # dplyr::arrange(desc(ClusterNegLog10P), ClusterSize) %>%
  # dplyr::group_by(dataset, Locus, SNP, Cell_type, Element) %>%
  # dplyr::slice_max(order_by = "count", n=1, with_ties = T) %>%
  # dplyr::slice_head() %>%
  dplyr::arrange(desc(Support)) %>%
  dplyr::mutate(Assay=gsub("promoters|enhancers","called_elements", Assay)) %>%
  subset(Support>2) %>%
  data.table::data.table()


createDT(regulatory_overlap)
# regulatory_overlap
```
 

# Session info

<details> 

```{r Session Info, attr.output='style="max-height: 400px;"'}
utils::sessionInfo()
```

</details> 
